### ISSUE TITLE
Different build behaviour for .net core web apps between local VS IDE and msbuild

### ISSUE THREAD

**Issue by vijayrkn** [dotnet/cli#10647]
_From @fromorbonia on January 10, 2019 9:26_

## Steps to reproduce
Publishing a .NET Core web application with a FileSystem profile under VS2017 IDE creates a directory with an EXE (e.g. MyProject.exe)
There is a Jenkins build server running msbuild, which correctly picks up the publish profile that was used and configured locally.
Place the resulting files on a server (with IIS) and a previosuly configured Web Site, and the site should start correctly.

## Expected  behavior
I would expect the build server to create exactly the same set of files, as produced by the local VS2017 publish process. 
Specifically, in this case, it should include MyProject.exe with a matching web.config file - so that once the files being copied to the IIS server the web site starts correctly.

## Actual behavior
The publish folder contains an apphost.exe and MyProject.dll, but not an EXE. The web.config points to the EXE that does not exist:
`<aspNetCore processPath=".\MyProject.exe"...`
The web site doesn't not run on IIS.

## Environment data
After some rounds of changes, I've ended up with the following key values (when looking at similar other posts people have asked about so have mentioned here) in the pubxml file:
```
    <UseAppHost>true</UseAppHost>
    <SelfContained>false</SelfContained>
    <_IsPortable>true</_IsPortable>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
```
I've only set UseAppHost to true, because it on the local VS IDE it creates the correct combination of web.config and files so that the web site starts correctly on deployment. I don't have a requirement to use or not use an apphost.exe file.

For the **local VS2017 IDE** this is:
.NET Core SDK (reflecting any global.json):
 Version:   2.1.401
 Commit:    91b1c13032

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.1.7601
 OS Platform: Windows
 RID:         win7-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.401\

Host (useful for support):
  Version: 2.1.3
  Commit:  124038c13e

.NET Core SDKs installed:
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

For the **Jenkins build** server:
12:53:26 .NET Core SDK (reflecting any global.json):
12:53:26  Version:   2.1.502
12:53:26  Commit:    c74ce8f29f
12:53:26 
12:53:26 Runtime Environment:
12:53:26  OS Name:     Windows
12:53:26  OS Version:  10.0.14393
12:53:26  OS Platform: Windows
12:53:26  RID:         win10-x64
12:53:26  Base Path:   C:\Program Files\dotnet\sdk\2.1.502\
12:53:26 
12:53:26 Host (useful for support):
12:53:26   Version: 2.1.6
12:53:26   Commit:  3f4f8eebd8
12:53:26 
12:53:26 .NET Core SDKs installed:
12:53:26   2.1.401 [C:\Program Files\dotnet\sdk]
12:53:26   2.1.500 [C:\Program Files\dotnet\sdk]
12:53:26   2.1.502 [C:\Program Files\dotnet\sdk]
12:53:26 
12:53:26 .NET Core runtimes installed:
12:53:26   Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
12:53:26   Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
12:53:26   Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
12:53:26   Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
12:53:26   Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
12:53:26   Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

_Copied from original issue: aspnet/websdk#470_

**Comment by vijayrkn** [comment]
_From @fromorbonia on January 14, 2019 16:59_

Does anyone have a view on this? Have I missed something obvious, or is there a real issue here? 
(@vijayrkn - I notice you've picked up a few of the posts, but the last 3 issues haven't been commented on by anyone). 
The build pipeline work I'm on is stalled at the moment because of this complication. 

Thanks.

**Comment by vijayrkn** [comment]
@fromorbonia - On the build servicer, does it produce the exe when you just run dotnet publish?

If you can share the output of the below command from both dev machine and build server then it would help with the investigation.

`dotnet publish /bl `

**Comment by vijayrkn** [comment]
_From @fromorbonia on January 15, 2019 15:37_

@vijayrkn - thanks for looking at this.

On the build server, with the command above (i.e. no explict publish profile), it does not produce the .exe.
If I run it with the PublishProfile switch:
`dotnet publish /bl "MySolution.sln" /p:PublishProfile=FolderProfile `
 The output includes apphost.exe (but does ignore the PublishUrl property in the profile).

In terms of the output, are you looking for a copy of msbuild.binlog? 

The publish directory output was:

```
15:20:17 01/15/2019  03:20 PM    <DIR>          .
15:20:17 01/15/2019  03:20 PM    <DIR>          ..
15:20:17 01/15/2019  03:19 PM               146 appsettings.Development.json
15:20:17 01/15/2019  03:19 PM               105 appsettings.json
15:20:17 06/06/2018  11:27 PM            52,728 dotnet-aspnet-codegenerator-design.dll
15:20:17 01/15/2019  03:19 PM                75 libman.json
15:20:17 04/30/2018  03:29 PM           689,584 Microsoft.CodeAnalysis.CSharp.Workspaces.dll
15:20:17 04/30/2018  03:29 PM         2,624,944 Microsoft.CodeAnalysis.Workspaces.dll
15:20:17 06/06/2018  11:27 PM            23,544 Microsoft.VisualStudio.Web.CodeGeneration.Contracts.dll
15:20:17 06/06/2018  11:26 PM            73,208 Microsoft.VisualStudio.Web.CodeGeneration.Core.dll
15:20:17 06/06/2018  11:26 PM            35,320 Microsoft.VisualStudio.Web.CodeGeneration.dll
15:20:17 06/06/2018  11:26 PM            69,104 Microsoft.VisualStudio.Web.CodeGeneration.EntityFrameworkCore.dll
15:20:17 06/06/2018  11:27 PM            28,656 Microsoft.VisualStudio.Web.CodeGeneration.Templating.dll
15:20:17 06/06/2018  11:27 PM            34,808 Microsoft.VisualStudio.Web.CodeGeneration.Utils.dll
15:20:17 06/06/2018  11:26 PM           157,688 Microsoft.VisualStudio.Web.CodeGenerators.Mvc.dll
15:20:17 04/24/2018  08:27 PM           108,688 NuGet.Frameworks.dll
15:20:17 01/15/2019  03:19 PM                 6 poolmaintenance.txt
15:20:17 01/15/2019  03:20 PM           247,223 MySolution.deps.json
15:20:17 01/15/2019  03:20 PM            12,288 MySolution.dll
15:20:17 01/15/2019  03:20 PM             2,240 MySolution.pdb
15:20:17 01/15/2019  03:20 PM               224 MySolution.runtimeconfig.json
15:20:17 01/15/2019  03:20 PM            75,776 MySolution.Views.dll
15:20:17 01/15/2019  03:20 PM             5,656 MySolution.Views.pdb
15:20:17 11/05/2016  04:55 AM            24,840 System.Composition.AttributedModel.dll
15:20:17 11/05/2016  04:55 AM            59,128 System.Composition.Convention.dll
15:20:17 11/05/2016  04:55 AM            62,184 System.Composition.Hosting.dll
15:20:17 11/05/2016  04:55 AM            29,928 System.Composition.Runtime.dll
15:20:17 11/05/2016  04:55 AM            64,760 System.Composition.TypedParts.dll
15:20:17 01/15/2019  03:20 PM               652 web.config
15:20:17 01/15/2019  03:19 PM               661 web.config2
15:20:17 01/15/2019  03:20 PM    <DIR>          wwwroot
```

**Comment by vijayrkn** [comment]
If you want to pass the publishprofile, then you need to use this commandline. Can you run this on both machines and share the binarylog output? This will help us determine why there is a change in behavior between the 2 machines. 

`msbuild MySolution.sln /p:DeployOnBuild=true /p:PublishProfile=FolderProfile /bl`

**Comment by vijayrkn** [comment]
_From @fromorbonia on January 15, 2019 16:18_

I did wonder if I could do that on the command line, will do that now on both.
The binlog file is 700KB+, is there a good place for me to post them? 
Thanks

**Comment by vijayrkn** [comment]
You can email it to me at vramak at microsoft

**Comment by vijayrkn** [comment]
@Peterhuene - Is there a bug in 2.1.502 that would prevent the exe from getting generated even when UseAppHost is set to true?  I will share the binlogs with you. UseAppHost is getting set in both cases.

/cc @livarcocc

**Comment by vijayrkn** [comment]
_From @fromorbonia on January 22, 2019 8:50_

@peterhuene - are you able to have a quick look at this? Thanks

**Comment by vijayrkn** [comment]
_From @peterhuene on January 22, 2019 17:26_

I'll take a look at the binlog and see what I can determine.  Sorry for the delay.

**Comment by vijayrkn** [comment]
_From @peterhuene on January 23, 2019 0:45_

@fromorbonia This is what I think is happening: various apphost related properties (e.g. `_DotNetAppHostExecutableName`) are set based on the RID in `Microsoft.NET.Sdk.targets`.  This happens prior to importing `Microsoft.NET.Sdk.Publish.targets`, which in turn is the file that imports your `pubxml` file.  The RID, however, changes as a result of the pubxml import.  This causes a problem as the restore is for the new RID (`win-x64` in this case), but we can't find the apphost because it's expected to not have an extension (`_NativeExecutableExtension` is empty).

A workaround would be to use `-r win-x64` for `dotnet publish` or pass `/p:RuntimeIdentifier=win-x64` if using MSBuild to force the RID to match before and after the pubxml is imported.

However, there's a separate bug I'm tracking (dotnet/cli#10602) where the publish operation copies `apphost.exe`, `hostpolicy.dll`, and `hostfxr.dll` to the output for framework-dependent ASP.NET applications; this prevents a framework-dependent deployed ASP.NET application from activating via the executable even when the correct executable is generated.

**Comment by vijayrkn** [comment]
_From @peterhuene on January 23, 2019 0:53_

@vijayrkn feel free to migrate this issue to `dotnet/cli` as this isn't a websdk issue.

**Comment by fromorbonia** [comment]
@peterhuene - using MSBuild no combination of RID setting whilst referencing the profile worked. I then tried some options with MSBuild and /p to pass in parameters, still could not get the .exe to build.
Switched to using dotnet publish, with -c Release -r win-x64 . This then generated a publish directory I could just deploy as is to IIS and the site then work. So I have a worksround now.

**Comment by peterhuene** [comment]
So I have a fix that corrects the following if `RuntimeIdentifier` is set in a publish profile:

* We now properly calculate the expected apphost filename for the runtime being targeted when resolving assets (e.g. if `RuntimeIdentifier` is `win-x64`, then we use `apphost.exe` instead of `apphost`).  Without this, we cannot find the unmodified apphost executable to customize.
* We now properly calculate `SelfContained` so that it also doesn't also need to be specified in the publish profile (a non-empty `RuntimeIdentifier` implies `SelfContained` is `true`).
* We now properly calculate `UseAppHost` so that it doesn't need to be specified in the publish profile (defaults to true for `netcoreapp2.2` or later published applications that specify a `RuntimeIdentifier`).
* We now properly calculate `DeployAppHost` so that it defaults to true when using the apphost.

The net result of these changes is that if you set a Windows`RuntimeIdentifier` and `SelfContained` to `false` in a publish profile, you will get the framework-dependent executable in the publish output.

I will attempt to propose these changes as a servicing fix for 2.2, but I'm cautiously optimistic regarding the fix's approval given the number of changes required to make this work properly.

**Comment by peterhuene** [comment]
Unfortunately my fix for respecting the expected default for `SelfContained` property from a publish profile that sets `RuntimeIdentifier` breaks self-contained publishing because `Microsoft.NETCore.App` has build property logic based on `SelfContained`.  This cannot be changed, so my approach must be rethought.

This is not unlike how the Test SDK changes the output type of the project after the evaluation of SDK properties.  I'd like to prevent having to duplicate the solution to that problem.

I might just have to limit the fix to publish profiles that set `RuntimeIdentifier`, `SelfContained`, and (possibly) `UseAppHost` explicitly.

**Comment by peterhuene** [comment]
After more consideration, the fix to this will require specifying `RuntimeIdentifier`, `SelfContained`, and `UseAppHost` in the publish profile (like we have in the example profile above).

It's going to be far too complex to properly default these values based on whatever important SDK properties are modified by restored packages.  Given that setting `RuntimeIdentifier` in a publish profile doesn't imply `SelfContained` today (i.e. not in any way a regression), users are likely already explicitly setting `SelfContained` to `true` if they wanted a self-contained publish.

I'll proceed with my limited-scope fix.

**Comment by fromorbonia** [comment]
@peterhuene - consistency is my main objective, so if consistency relies on publishing a profile, then that works for me. This also removes any ambiguity when changing .net core versions in the future. Thanks.

**Comment by peterhuene** [comment]
This is now fixed for 3.0 with dotnet/sdk#3183.

We'll have to check if both the Web SDK fix and the .NET Core SDK fix meet the servicing bar to backport.

**Comment by peterhuene** [comment]
@fromorbonia I don't believe the fix we made for 3.0 can be ported to a servicing release as it requires a coordinated change from both the .NET Core SDK and the Web SDK to make work.  Combined with the fact that this isn't a regression from previous versions and there is a simple-enough workaround, it won't meet the servicing bar to backport to the 2.x versions.

With a "go live" preview of the 3.0 SDK coming soon, I hope you find it fixes this issue.  Until then, I would use the workaround described in my comments above.  As this is fixed in 3.0, I'm going to close this issue for now.  Thanks again for reporting it to us.
